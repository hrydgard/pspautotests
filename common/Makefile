TARGET_LIB = libcommon.a
OBJS = common.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak
