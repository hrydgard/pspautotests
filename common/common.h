#ifndef _COMMON_H
#define _COMMON_H 1

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <pspctrl.h>

#undef main
#define main test_main

extern int RUNNING_ON_EMULATOR;

void emulatorEmitScreenshot();
void emulatorSendSceCtrlData(SceCtrlData* pad_data);

/*
void emitInt(int v);
void emitFloat(float v);
void emitString(char *v);
void emitComment(char *v);
void emitMemoryBlock(void *address, unsigned int size);
void emitHex(void *address, unsigned int size);
#define emitStringf(format, ...) { char temp[1024]; sprintf(temp, format, __VA_ARGS__); emitString(temp); }
*/

#endif