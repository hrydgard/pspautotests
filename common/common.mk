# Include this to simplify your test include file.
#
# Usage:
#
# TARGETS=mytest
# include ../../path/to/common/common.mk

BUILD_PRX = 1
USE_PSPSDK_LIBC = 1
PSP_FW_VERSION = 500

INCDIR := $(INCDIR) . $(COMMON_DIR)
LIBDIR := $(LIBDIR) . $(COMMON_DIR)

ifndef CFLAGS
CFLAGS = -g -G0 -Wall -O0 -fno-strict-aliasing
endif
ifndef CXXFLAGS
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti 
endif
ifndef ASFLAGS
ASFLAGS = $(CFLAGS)
endif
ifndef LDFLAGS
LDFLAGS = -G0
endif

ifndef LIBS
LIBS = -lpspgu -lpsprtc -lpspctrl -lpspmath -lcommon -lc -lm
endif

TARGET = $(firstword $(TARGETS))
OBJS = $(firstword $(TARGETS)).o

PSPSDK = $(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.elf: %.o $(EXTRA_OBJS) $(EXPORT_OBJ)
	$(LINK.c) $^ $(LIBS) -o $@
	$(FIXUP) $@

%.prx: %.elf
	psp-prxgen $< $@

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

all: $(TARGETS:=.prx)
clean: EXTRA_TARGETS:=$(EXTRA_TARGETS) $(TARGETS:=.prx)
