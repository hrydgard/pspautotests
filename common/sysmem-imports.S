.set noreorder

#include "pspimport.s"

IMPORT_START "SysMemUserForUser",0x40000000

IMPORT_FUNC  "SysMemUserForUser",0xA291F107,sceKernelMaxFreeMemSize
IMPORT_FUNC  "SysMemUserForUser",0xF919F628,sceKernelTotalFreeMemSize
IMPORT_FUNC  "SysMemUserForUser",0x2A3E5280,sceKernelQueryMemoryInfo

IMPORT_FUNC  "SysMemUserForUser",0x3FC9AE6A,sceKernelDevkitVersion
IMPORT_FUNC  "SysMemUserForUser",0x13A5ABEF,sceKernelPrintf
IMPORT_FUNC  "SysMemUserForUser",0x39F49610,sceKernelGetPTRIG
IMPORT_FUNC  "SysMemUserForUser",0x6231A71D,sceKernelSetPTRIG

IMPORT_FUNC  "SysMemUserForUser",0x237DBD4F,sceKernelAllocPartitionMemory
IMPORT_FUNC  "SysMemUserForUser",0xB6D61D02,sceKernelFreePartitionMemory
IMPORT_FUNC  "SysMemUserForUser",0x9D9A5BA1,sceKernelGetBlockHeadAddr

IMPORT_FUNC  "SysMemUserForUser",0xDB83A952,sceKernelGetMemoryBlockPtr
IMPORT_FUNC  "SysMemUserForUser",0xFE707FDF,sceKernelAllocMemoryBlock
IMPORT_FUNC  "SysMemUserForUser",0x50F61D8A,sceKernelFreeMemoryBlock

IMPORT_FUNC  "SysMemUserForUser",0xF77D77CB,sceKernelSetCompilerVersion
IMPORT_FUNC  "SysMemUserForUser",0xFC114573,sceKernelGetCompiledSdkVersion
IMPORT_FUNC  "SysMemUserForUser",0x7591C7DB,sceKernelSetCompiledSdkVersion
IMPORT_FUNC  "SysMemUserForUser",0x342061E5,sceKernelSetCompiledSdkVersion370
IMPORT_FUNC  "SysMemUserForUser",0x315AD3A0,sceKernelSetCompiledSdkVersion380_390
IMPORT_FUNC  "SysMemUserForUser",0xEBD5C3E6,sceKernelSetCompiledSdkVersion395
IMPORT_FUNC  "SysMemUserForUser",0x057E7380,sceKernelSetCompiledSdkVersion401_402
IMPORT_FUNC  "SysMemUserForUser",0x91DE343C,sceKernelSetCompiledSdkVersion500_505
IMPORT_FUNC  "SysMemUserForUser",0x7893F79A,sceKernelSetCompiledSdkVersion507
IMPORT_FUNC  "SysMemUserForUser",0x35669D4C,sceKernelSetCompiledSdkVersion600_602
IMPORT_FUNC  "SysMemUserForUser",0x1B4217BC,sceKernelSetCompiledSdkVersion603_605
IMPORT_FUNC  "SysMemUserForUser",0x358CA1BB,sceKernelSetCompiledSdkVersion606

# Not sure if correct.
IMPORT_FUNC  "SysMemUserForUser",0xA6848DF8,sceKernelSetUsersystemLibWork
IMPORT_FUNC  "SysMemUserForUser",0xACBD88CA,sceKernelTotalMemSize
IMPORT_FUNC  "SysMemUserForUser",0xD8DE5C1E,SysMemUserForUser_D8DE5C1E
IMPORT_FUNC  "SysMemUserForUser",0x945E45DA,SysMemUserForUser_945E45DA