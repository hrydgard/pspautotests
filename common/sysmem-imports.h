extern int sceKernelSetCompiledSdkVersion(u32 ver);
extern int sceKernelSetCompiledSdkVersion370(u32 ver);
extern int sceKernelSetCompiledSdkVersion380_390(u32 ver);
extern int sceKernelSetCompiledSdkVersion395(u32 ver);
extern int sceKernelSetCompiledSdkVersion401_402(u32 ver);
extern int sceKernelSetCompiledSdkVersion500_505(u32 ver);
extern int sceKernelSetCompiledSdkVersion507(u32 ver);
extern int sceKernelSetCompiledSdkVersion600_602(u32 ver);
extern int sceKernelSetCompiledSdkVersion603_605(u32 ver);
extern int sceKernelSetCompiledSdkVersion606(u32 ver);
extern u32 sceKernelGetCompiledSdkVersion();