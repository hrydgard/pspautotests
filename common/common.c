#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <assert.h>
#include <pspdebug.h>
#include <pspthreadman.h>
#include <psploadexec.h>
#include <pspctrl.h>
#include <pspdisplay.h>
#include <pspiofilemgr.h>
//#include <pspkdebug.h>

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <sys/lock.h>
#include <sys/fcntl.h>

//#include "local.h"

#include "sysmem-imports.h"

/*
enum PspThreadAttributes
{
	PSP_THREAD_ATTR_VFPU = 0x00004000,
	PSP_THREAD_ATTR_USER = 0x80000000,
	PSP_THREAD_ATTR_USBWLAN = 0xa0000000,
	PSP_THREAD_ATTR_VSH = 0xc0000000,
	PSP_THREAD_ATTR_SCRATCH_SRAM = 0x00008000,
	PSP_THREAD_ATTR_NO_FILLSTACK = 0x00100000,
	PSP_THREAD_ATTR_CLEAR_STACK = 0x00200000,
};
*/
/*
enum PspModuleInfoAttr
{
	PSP_MODULE_USER			= 0,
	PSP_MODULE_NO_STOP		= 0x0001,
	PSP_MODULE_SINGLE_LOAD	= 0x0002,
	PSP_MODULE_SINGLE_START	= 0x0004,
	PSP_MODULE_KERNEL		= 0x1000,
};
*/

PSP_MODULE_INFO("TESTMODULE", PSP_MODULE_USER, 1, 0);
PSP_MAIN_THREAD_ATTR(PSP_THREAD_ATTR_USER | PSP_THREAD_ATTR_VFPU);
//PSP_MAIN_THREAD_ATTR(PSP_THREAD_ATTR_USER);

#define EMULATOR_DEVCTL__GET_HAS_DISPLAY 0x00000001
#define EMULATOR_DEVCTL__SEND_OUTPUT     0x00000002
#define EMULATOR_DEVCTL__IS_EMULATOR     0x00000003
#define EMULATOR_DEVCTL__SEND_CTRLDATA   0x00000010
#define EMULATOR_DEVCTL__EMIT_SCREENSHOT 0x00000020

unsigned int RUNNING_ON_EMULATOR = 0;
unsigned int CHECKPOINT_ENABLE_TIME = 0;
unsigned int CHECKPOINT_OUTPUT_DIRECT = 0;
unsigned int HAS_DISPLAY = 1;

// 21 MB to give space for thread stacks and etc.
unsigned int sce_newlib_heap_kb_size = 21504;

extern int test_main(int argc, char *argv[]);

FILE stdout_back = {NULL};
//int KprintfFd = 0;

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	if (CHECKPOINT_OUTPUT_DIRECT) {
		// This is easier to debug in the emulator, but printf() reschedules on the real PSP.
		vprintf(format, args);
	} else {
		schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	}
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

SceUID reschedThread;
volatile int didResched = 0;
int reschedFunc(SceSize argc, void *argp) {
	didResched = 1;
	return 0;
}

u64 lastCheckpoint = 0;
void checkpoint(const char *format, ...) {
	u64 currentCheckpoint = sceKernelGetSystemTimeWide();
	if (CHECKPOINT_ENABLE_TIME) {
		schedf("[%s/%lld] ", didResched ? "r" : "x", currentCheckpoint - lastCheckpoint);
	} else {
		schedf("[%s] ", didResched ? "r" : "x");
	}

	sceKernelTerminateThread(reschedThread);

	if (format != NULL) {
		va_list args;
		va_start(args, format);
		if (CHECKPOINT_OUTPUT_DIRECT) {
			// This is easier to debug in the emulator, but printf() reschedules on the real PSP.
			vprintf(format, args);
		} else {
			schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
		}
		va_end(args);
	}

	didResched = 0;
	sceKernelStartThread(reschedThread, 0, NULL);

	if (format != NULL) {
		schedf("\n");
	}
	
	lastCheckpoint = currentCheckpoint;
}

void checkpointNext(const char *title) {
	if (schedfBufferPos != 0) {
		schedf("\n");
	}
	flushschedf();
	didResched = 0;
	if (title != NULL)
		checkpoint(title);
}

static int writeStdoutHook(struct _reent *ptr, void *cookie, const char *buf, int buf_len) {
	char temp[1024 + 1];

	//if (KprintfFd > 0) sceIoWrite(KprintfFd, buf, buf_len);
	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("emulator:", EMULATOR_DEVCTL__SEND_OUTPUT, (void *)buf, buf_len, NULL, 0);
	}

	if (buf_len < sizeof(temp)) {
		if (HAS_DISPLAY) {
			memcpy(temp, buf, buf_len);
			temp[buf_len] = 0;

			//Kprintf("%s", temp);
			pspDebugScreenPrintf("%s", temp);
		}
	}
	
	if (stdout_back._write != NULL) {
		return stdout_back._write(ptr, cookie, buf, buf_len);
	} else {
		return buf_len;
	}
}

typedef int (*SdkVerFunc)(u32 ver);
typedef struct SdkVerFuncTable {
	u32 id;
	SdkVerFunc func;
} SdkVerFuncTable;

static SdkVerFuncTable sdkVerFuncs[] = {
	{0, sceKernelSetCompiledSdkVersion},
	{370, sceKernelSetCompiledSdkVersion370},
	{380, sceKernelSetCompiledSdkVersion380_390},
	{395, sceKernelSetCompiledSdkVersion395},
	{401, sceKernelSetCompiledSdkVersion401_402},
	{500, sceKernelSetCompiledSdkVersion500_505},
	{507, sceKernelSetCompiledSdkVersion507},
	{600, sceKernelSetCompiledSdkVersion600_602},
	{603, sceKernelSetCompiledSdkVersion603_605},
	{606, sceKernelSetCompiledSdkVersion606},
};

static void updateSdkVer(int argc, char *argv[]) {
	int i = 0;
	u32 ver = 0xFFFFFFFF;
	u32 funcID = 0;
	for (i = 1; i < argc; ++i) {
		if (!strncmp(argv[i], "--sdkver=", strlen("--sdkver="))) {
			ver = strtoul(argv[i] + strlen("--sdkver="), NULL, 16);
		}
		if (!strncmp(argv[i], "--sdkver-func=", strlen("--sdkver-func="))) {
			funcID = strtol(argv[i] + strlen("--sdkver-func="), NULL, 10);
		}
	}

	SdkVerFunc func = NULL;
	for (i = 0; i < sizeof(sdkVerFuncs) / sizeof(sdkVerFuncs[0]); ++i) {
		if (sdkVerFuncs[i].id == funcID) {
			func = sdkVerFuncs[i].func;
		}
	}

	if (func == NULL) {
		fprintf(stderr, "Unknown sdkver-func value.\n");
		exit(1);
	}

	if (ver != 0xFFFFFFFF) {
		if (func(ver) != 0) {
			printf("WARNING: Setting sdkver returned failure.\n");
		}
	}
}

void test_begin() {
	if (HAS_DISPLAY) {
		pspDebugScreenInit();
	}

	if (RUNNING_ON_EMULATOR && !HAS_DISPLAY) {
		fclose(stdout);
		stdout = fmemopen(alloca(4), 4, "wb");
		stdout_back._write = NULL;
		
		//stderr = stdout;
		setbuf(stdout, NULL);
	} else {
    // Send the output to the host.
		freopen("host0:/__testoutput.txt", "wb", stdout);
		freopen("host0:/__testerror.txt", "wb", stderr);
		stdout_back._write = stdout->_write;
	}
	stdout->_write = writeStdoutHook;

	setvbuf(stdout, NULL, _IONBF, 0);
	setvbuf(stderr, NULL, _IONBF, 0);
	
	setbuf(stderr, NULL);

	reschedThread = sceKernelCreateThread("resched", &reschedFunc, sceKernelGetThreadCurrentPriority(), 0x1000, 0, NULL);
}

void test_end() {
	flushschedf();

	fflush(stdout);
	fflush(stderr);
	
	fclose(stdout);
	fclose(stderr);

	if (!RUNNING_ON_EMULATOR) {
		FILE *finish = fopen("host0:/__testfinish.txt", "wb");
		if (finish)
		{
			fwrite("1", sizeof(char), 1, finish);
			fclose(finish);
		}
	}

  // Disabled the wait, much more convienent when running automated.
	if (0 && !RUNNING_ON_EMULATOR) {
		SceCtrlData key;
		while (1) {
			sceCtrlReadBufferPositive(&key, 1);
			if (key.Buttons & PSP_CTRL_CROSS) break;
		}
	}
	
	//fclose(stdout);
	sceKernelExitGame();
	
	exit(0);
}

int test_psp_exit_callback(int arg1, int arg2, void *common) {
	exit(0);
	return 0;
}

int test_psp_callback_thread(SceSize args, void *argp) {
	int cbid;
	cbid = sceKernelCreateCallback("Exit Callback", test_psp_exit_callback, NULL);
	sceKernelRegisterExitCallback(cbid);
	sceKernelSleepThreadCB();
	return 0;
}

int test_psp_setup_callbacks(void) {
	int thid = 0;
	thid = sceKernelCreateThread("update_thread",  test_psp_callback_thread, 0x11, 0xFA0, 0, 0);
	if (thid >= 0) sceKernelStartThread(thid, 0, 0);
	return thid;
}

//#define START_WITH "ms0:/PSP/GAME/virtual"

/*
void emitInt(int v) {
	asm("syscall 0x1010");
}

void emitFloat(float v) {
	asm("syscall 0x1011");
}

void emitString(char *v) {
	asm("syscall 0x1012");
}

void emitComment(char *v) {
	asm("syscall 0x1012");
}

void emitMemoryBlock(void *address, unsigned int size) {
	asm("syscall 0x1013");
}

void emitHex(void *address, unsigned int size) {
	asm("syscall 0x1014");
}
*/

unsigned char bmpHeader[54] = {
	0x42, 0x4D, 0x38, 0x80, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x36, 0x00,
	0x00, 0x00, 0x28, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x10, 0x01,
	0x00, 0x00, 0x01, 0x00, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x80,
	0x08, 0x00, 0x12, 0x0B, 0x00, 0x00, 0x12, 0x0B, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00
};


uint extractBits(uint value, int offset, int size) {
	return (value >> offset) & ((1 << size) - 1);
}

void emulatorEmitScreenshot() {
	int file;

	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("kemulator:", EMULATOR_DEVCTL__EMIT_SCREENSHOT, NULL, 0, NULL, 0);
	}
	else
	{
		uint topaddr;
		int bufferwidth;
		int pixelformat;

		sceDisplayGetFrameBuf((void **)&topaddr, &bufferwidth, &pixelformat, 0);
		
        if (topaddr & 0x80000000) {
            topaddr |= 0xA0000000;
        } else {
            topaddr |= 0x40000000;
        }
	
		if ((file = sceIoOpen("host0:/__screenshot.bmp", PSP_O_CREAT | PSP_O_WRONLY | PSP_O_TRUNC, 0777)) >= 0) {
			int y, x;
			uint c;
			uint* vram_row;
			uint* row_buf = (uint *)malloc(512 * 4);
			sceIoWrite(file, &bmpHeader, sizeof(bmpHeader));
			for (y = 0; y < 272; y++) {
				vram_row = (uint *)(topaddr + 512 * 4 * (271 - y));
				for (x = 0; x < 512; x++) {
					c = vram_row[x];
					/*
					row_buf[x] = (
						((extractBits(c,  0, 8)) <<  0) |
						((extractBits(c,  8, 8)) <<  8) |
						((extractBits(c, 16, 8)) << 16) |
						((                0x00 ) << 24) |
					0);
					*/
					row_buf[x] = (
						((extractBits(c, 16, 8)) <<  0) |
						((extractBits(c,  8, 8)) <<  8) |
						((extractBits(c,  0, 8)) << 16) |
						((                0x00 ) << 24) |
					0);
				}
				sceIoWrite(file, row_buf, 512 * 4);
			}
			free(row_buf);
			//sceIoWrite(file, (void *)topaddr, bufferwidth * 272 * 4);
			//sceIoFlush();
			sceIoClose(file);
		}
	}
}

void emulatorSendSceCtrlData(SceCtrlData* pad_data) {
	sceIoDevctl("kemulator:", EMULATOR_DEVCTL__SEND_CTRLDATA, pad_data, sizeof(SceCtrlData), NULL, 0);
}

int main(int argc, char *argv[]) {
	int retval = 0;

	//KprintfFd = sceIoOpen("emulator:/Kprintf", O_WRONLY, 0777);
	//RUNNING_ON_EMULATOR = (KprintfFd > 0);
	
	if (sceIoDevctl("kemulator:", EMULATOR_DEVCTL__IS_EMULATOR, NULL, 0, NULL, 0) == 0) {
		RUNNING_ON_EMULATOR = 1;
	}
	
	// TEMP HACK
	//RUNNING_ON_EMULATOR = 0;

	if (RUNNING_ON_EMULATOR) {
		sceIoDevctl("kemulator:", EMULATOR_DEVCTL__GET_HAS_DISPLAY, NULL, 0, &HAS_DISPLAY, sizeof(HAS_DISPLAY));
	}
	
	#ifdef GRAPHIC_TEST
		HAS_DISPLAY = 0;
	#endif

	//if (strncmp(argv[0], START_WITH, strlen(START_WITH)) == 0) RUNNING_ON_EMULATOR = 1;

	if (!RUNNING_ON_EMULATOR) {
		test_psp_setup_callbacks();
	}
	atexit(sceKernelExitGame);

	test_begin();
	{
		pspDebugScreenPrintf("RUNNING_ON_EMULATOR: %s - %s\n", RUNNING_ON_EMULATOR ? "yes" : "no", argv[0]);
		updateSdkVer(argc, argv);

		retval = test_main(argc, argv);
	}
	test_end();
	
	return retval;
}
