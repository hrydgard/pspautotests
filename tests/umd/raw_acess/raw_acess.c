#include <common.h>

#include <pspsdk.h>
#include <pspkernel.h>
#include <pspthreadman.h>
#include <psploadexec.h>
#include <pspumd.h>
#include <pspiofilemgr.h>

char data1[472] = {0};
char data2[472] = {0};
char data3[472] = {0};

unsigned char dataRef[472] =
{
	0x00, 0x50, 0x53, 0x46, 0x01, 0x01, 0x00, 0x00, 0xB4, 0x00, 0x00, 0x00,
	0x20, 0x01, 0x00, 0x00, 0x0A, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x04,
	0x04, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x09, 0x00, 0x04, 0x02, 0x03, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00,
	0x04, 0x00, 0x00, 0x00, 0x12, 0x00, 0x04, 0x02, 0x0A, 0x00, 0x00, 0x00,
	0x10, 0x00, 0x00, 0x00, 0x08, 0x00, 0x00, 0x00, 0x1A, 0x00, 0x04, 0x04,
	0x04, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x18, 0x00, 0x00, 0x00,
	0x26, 0x00, 0x04, 0x04, 0x04, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00,
	0x1C, 0x00, 0x00, 0x00, 0x31, 0x00, 0x04, 0x02, 0x05, 0x00, 0x00, 0x00,
	0x08, 0x00, 0x00, 0x00, 0x20, 0x00, 0x00, 0x00, 0x3E, 0x00, 0x04, 0x04,
	0x04, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x28, 0x00, 0x00, 0x00,
	0x4D, 0x00, 0x04, 0x02, 0x05, 0x00, 0x00, 0x00, 0x08, 0x00, 0x00, 0x00,
	0x2C, 0x00, 0x00, 0x00, 0x5C, 0x00, 0x04, 0x04, 0x04, 0x00, 0x00, 0x00,
	0x04, 0x00, 0x00, 0x00, 0x34, 0x00, 0x00, 0x00, 0x63, 0x00, 0x04, 0x02,
	0x0C, 0x00, 0x00, 0x00, 0x80, 0x00, 0x00, 0x00, 0x38, 0x00, 0x00, 0x00,
	0x42, 0x4F, 0x4F, 0x54, 0x41, 0x42, 0x4C, 0x45, 0x00, 0x43, 0x41, 0x54,
	0x45, 0x47, 0x4F, 0x52, 0x59, 0x00, 0x44, 0x49, 0x53, 0x43, 0x5F, 0x49,
	0x44, 0x00, 0x44, 0x49, 0x53, 0x43, 0x5F, 0x4E, 0x55, 0x4D, 0x42, 0x45,
	0x52, 0x00, 0x44, 0x49, 0x53, 0x43, 0x5F, 0x54, 0x4F, 0x54, 0x41, 0x4C,
	0x00, 0x44, 0x49, 0x53, 0x43, 0x5F, 0x56, 0x45, 0x52, 0x53, 0x49, 0x4F,
	0x4E, 0x00, 0x50, 0x41, 0x52, 0x45, 0x4E, 0x54, 0x41, 0x4C, 0x5F, 0x4C,
	0x45, 0x56, 0x45, 0x4C, 0x00, 0x50, 0x53, 0x50, 0x5F, 0x53, 0x59, 0x53,
	0x54, 0x45, 0x4D, 0x5F, 0x56, 0x45, 0x52, 0x00, 0x52, 0x45, 0x47, 0x49,
	0x4F, 0x4E, 0x00, 0x54, 0x49, 0x54, 0x4C, 0x45, 0x00, 0x00, 0x00, 0x00,
	0x01, 0x00, 0x00, 0x00, 0x55, 0x47, 0x00, 0x00, 0x55, 0x43, 0x4A, 0x53,
	0x31, 0x30, 0x30, 0x34, 0x31, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x31, 0x2E, 0x30, 0x30,
	0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x31, 0x2E, 0x35, 0x30,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x00, 0x00, 0x43, 0x75, 0x62, 0x65,
	0x20, 0x73, 0x61, 0x6D, 0x70, 0x6C, 0x65, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00
};


#define FILE_NAME "disc0:/PSP_GAME/PARAM.SFO"
#define UMD0_FILE_NAME "umd0:/PSP_GAME/PARAM.SFO"

int main(int argc, char **argv)
{
	char temp[1024];
	SceIoStat sceIoStat;
	int fileHandle;

	printf("sceIoGetstat: %d\n", sceIoGetstat(FILE_NAME, &sceIoStat));
	printf("Size: %lld\n", sceIoStat.st_size);
	printf("LBN: %d\n", sceIoStat.st_private[0]);
	// TEST normal file access.
	{
		fileHandle = sceIoOpen(FILE_NAME, PSP_O_RDONLY, 0777);
		printf("HandleNormal: %s\n", (fileHandle > 0) ? "Ok" : "Error");
		sceIoRead(fileHandle, data1, sizeof(data1));
		sceIoClose(fileHandle);
		printf("memcmp_Normal: %d\n", memcmp(data1, dataRef, sizeof(dataRef)));
	}
	// TEST sce_lbn access
	{
		sprintf(temp, "disc0:/sce_lbn0x%08X_size0x%08X", (int)sceIoStat.st_private[0], (int)sceIoStat.st_size);
		fileHandle = sceIoOpen(temp, PSP_O_RDONLY, 0777);
		printf("HandleRaw: %s\n", (fileHandle > 0) ? "Ok" : "Error");
		sceIoRead(fileHandle, data2, sizeof(data2));
		sceIoClose(fileHandle);
		printf("memcmp_Raw: %d\n", memcmp(data2, dataRef, sizeof(dataRef)));
	}
	// TEST umd0:/ access
	{
		fileHandle = sceIoOpen(UMD0_FILE_NAME, PSP_O_RDONLY, 0777);
		printf("HandleUmd0: %s\n", (fileHandle > 0) ? "Ok" : "Error");
		sceIoLseek(fileHandle, (long long int)sceIoStat.st_private[0] * 0x800, SEEK_SET);
		sceIoRead(fileHandle, data3, sizeof(data3));
		sceIoClose(fileHandle);
		printf("memcmp_Umd0: %d\n", memcmp(data3, dataRef, sizeof(dataRef)));
	}

	return 0;
}