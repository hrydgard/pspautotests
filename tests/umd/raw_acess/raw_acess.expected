sceIoGetstat: 0
Size: 472
LBN: 26
HandleNormal: Ok
memcmp_Normal: 0
HandleRaw: Ok
memcmp_Raw: 0
HandleUmd0: Ok
memcmp_Umd0: 0
